File: workers/common/schema.py
"""
Shared data schemas for all SituationRoom workers.

Think of these as "blueprints" — they define exactly what fields each piece of
data must have. Pydantic (the library used here) validates the data automatically:
if a required field is missing or the wrong type, it raises an error immediately
so bugs are caught early rather than silently corrupting the database.

Two categories of data:
  Category A — moving entities streamed live (aircraft, ships)
  Category B — static/derived layers fetched on demand (events, jamming, TLEs)
"""

from __future__ import annotations

from datetime import datetime, date as DateType
from typing import Any, Literal, Optional
from uuid import UUID

from pydantic import BaseModel, Field


# ---------------------------------------------------------------------------
# Category A — NormalizedEntity
# ---------------------------------------------------------------------------
# Every moving thing on the map (aircraft, ship) is normalized into this shape
# before being published to Redis and inserted into entity_positions.
#
# The "id" field follows the pattern "{source}:{identifier}", e.g.:
#   "adsb:abc123"   — an aircraft identified by its ICAO24 hex code
#   "ais:987654321" — a ship identified by its MMSI number

class NormalizedEntity(BaseModel):
    id: str = Field(
        description='Globally unique entity id in the form "{source}:{identifier}"'
    )
    source: Literal["adsb", "ais", "satellite"] = Field(
        description="Which data feed this came from"
    )
    entity_type: Literal["aircraft", "ship", "satellite"] = Field(
        description=(
            "What kind of real-world object this is. "
            "'satellite' is used in Step 9 when the client-side Web Worker "
            "propagates TLE positions and injects them directly into the "
            "Zustand store — the Python workers never produce satellite entities."
        )
    )
    lat: float = Field(description="Latitude in decimal degrees (-90 to 90)")
    lon: float = Field(description="Longitude in decimal degrees (-180 to 180)")
    alt_m: Optional[float] = Field(
        default=None, description="Altitude in metres above sea level (aircraft only)"
    )
    heading_deg: Optional[float] = Field(
        default=None, description="Direction of travel in degrees (0 = North, clockwise)"
    )
    speed_knots: Optional[float] = Field(
        default=None, description="Speed in nautical miles per hour"
    )
    callsign: Optional[str] = Field(
        default=None,
        description="Flight number (aircraft) or vessel name (ship)"
    )
    metadata: dict[str, Any] = Field(
        default_factory=dict,
        description="Source-specific extras, e.g. {'icao24': 'abc123', 'country': 'US'}"
    )
    timestamp: datetime = Field(
        description="UTC time when this position was recorded"
    )


# ---------------------------------------------------------------------------
# Category B — Event
# ---------------------------------------------------------------------------
# A manually pinned incident on the map: strike, explosion, military movement, etc.
# Created by an admin via POST /api/events.

class Event(BaseModel):
    id: Optional[UUID] = Field(
        default=None,
        description="Database UUID — None before the row is inserted"
    )
    created_at: Optional[datetime] = Field(
        default=None, description="When this event was added to the system (set by DB)"
    )
    event_time: datetime = Field(
        description="When the real-world incident occurred (UTC)"
    )
    lat: float = Field(description="Latitude where the event happened")
    lon: float = Field(description="Longitude where the event happened")
    event_type: str = Field(
        description="Category: 'strike' | 'explosion' | 'military_movement' | 'protest' | 'other'"
    )
    title: str = Field(description="Short description, e.g. 'Missile strike on fuel depot'")
    description: Optional[str] = Field(default=None, description="Longer details")
    source_url: Optional[str] = Field(
        default=None, description="Link to a news article or source"
    )
    verified: bool = Field(
        default=False, description="True if an admin has confirmed this event"
    )
    contributor: Optional[str] = Field(
        default=None, description="Who submitted it (for attribution)"
    )


# ---------------------------------------------------------------------------
# Category B — JammingHex
# ---------------------------------------------------------------------------
# One row in the GPSJam daily dataset.
# Each hex represents a geographic hexagon (H3 format) and the percentage of
# aircraft over it that reported GPS interference.

class JammingHex(BaseModel):
    h3_index: str = Field(
        description="H3 hexagon identifier, e.g. '8928308280fffff'"
    )
    interference_pct: float = Field(
        description="0.0–100.0: percentage of flights reporting GPS issues in this hex"
    )
    date: DateType = Field(description="The calendar day this reading covers")


# ---------------------------------------------------------------------------
# Category B — TLERecord
# ---------------------------------------------------------------------------
# One satellite's Two-Line Element set, fetched from CelesTrak.
# The frontend uses these to compute satellite positions client-side.

class TLERecord(BaseModel):
    norad_cat_id: int = Field(
        description="NORAD catalog number — the universal unique ID for every satellite"
    )
    name: str = Field(description="Human-readable name, e.g. 'STARLINK-1234'")
    tle_line1: str = Field(description="First line of the TLE data string")
    tle_line2: str = Field(description="Second line of the TLE data string")
    constellation: str = Field(
        description="Group: 'starlink' | 'iridium-next' | 'active' | etc."
    )
    fetched_at: Optional[datetime] = Field(
        default=None,
        description="When we last retrieved this TLE from CelesTrak (set by DB)"
    )

File: api/pkg/schema/schema.go
// Package schema defines the shared data structures used across the Go API.
//
// These are the Go equivalents of the Python Pydantic models in
// workers/common/schema.py and the TypeScript interfaces in
// frontend/src/types/.
//
// Two categories:
//   - Category A: NormalizedEntity — moving things streamed via WebSocket
//   - Category B: Event, JammingHex, TLERecord — static layers served via REST
//
// The JSON field names here must match what the frontend TypeScript types expect.
package schema

import (
	"time"

	"github.com/google/uuid"
)

// ---------------------------------------------------------------------------
// Category A — NormalizedEntity
// ---------------------------------------------------------------------------
// A moving object on the map (aircraft or ship), normalized from its source
// feed (OpenSky for ADS-B, AISStream for ships) into a common shape.
//
// Redis messages use this same struct, serialised as JSON, inside an envelope:
//
//	{ "type": "update", "entity": { ...NormalizedEntity fields... } }
//	{ "type": "remove", "id": "adsb:abc123" }

// NormalizedEntity is the canonical representation of a tracked moving entity.
type NormalizedEntity struct {
	// ID is "{source}:{identifier}", e.g. "adsb:abc123" or "ais:987654321"
	ID         string  `json:"id"`
	Source     string  `json:"source"`      // "adsb" | "ais"
	EntityType string  `json:"entity_type"` // "aircraft" | "ship"
	Lat        float64 `json:"lat"`
	Lon        float64 `json:"lon"`

	// Optional fields — pointer so they serialise as null when absent
	AltM       *float64 `json:"alt_m,omitempty"`
	HeadingDeg *float64 `json:"heading_deg,omitempty"`
	SpeedKnots *float64 `json:"speed_knots,omitempty"`
	Callsign   *string  `json:"callsign,omitempty"`

	// Metadata holds source-specific extras (ICAO24, MMSI, country, etc.)
	Metadata  map[string]any `json:"metadata"`
	Timestamp time.Time      `json:"timestamp"`
}

// RedisMessage is the envelope that wraps entity updates on Redis Pub/Sub.
// Workers publish this; the Go gateway reads it and fans out to WebSocket clients.
type RedisMessage struct {
	// Type is "update" (entity position changed) or "remove" (entity gone)
	Type string `json:"type"`

	// Entity is set when Type == "update"
	Entity *NormalizedEntity `json:"entity,omitempty"`

	// ID is set when Type == "remove", e.g. "adsb:abc123"
	ID string `json:"id,omitempty"`
}

// ---------------------------------------------------------------------------
// Category B — Event
// ---------------------------------------------------------------------------
// A manually pinned incident (strike, explosion, military movement, etc.)
// Served via GET /api/events; created via POST /api/events (admin only).

// Event represents a single user-contributed map incident.
type Event struct {
	ID          uuid.UUID  `json:"id"`
	CreatedAt   time.Time  `json:"created_at"`
	EventTime   time.Time  `json:"event_time"`
	Lat         float64    `json:"lat"`
	Lon         float64    `json:"lon"`
	EventType   string     `json:"event_type"` // "strike" | "explosion" | "military_movement" | "protest" | "other"
	Title       string     `json:"title"`
	Description *string    `json:"description,omitempty"`
	SourceURL   *string    `json:"source_url,omitempty"`
	Verified    bool       `json:"verified"`
	Contributor *string    `json:"contributor,omitempty"`
}

// CreateEventRequest is the body accepted by POST /api/events.
// It omits server-assigned fields (id, created_at, verified).
type CreateEventRequest struct {
	EventTime   time.Time `json:"event_time"`
	Lat         float64   `json:"lat"`
	Lon         float64   `json:"lon"`
	EventType   string    `json:"event_type"`
	Title       string    `json:"title"`
	Description *string   `json:"description,omitempty"`
	SourceURL   *string   `json:"source_url,omitempty"`
	Contributor *string   `json:"contributor,omitempty"`
}

// ---------------------------------------------------------------------------
// Category B — JammingHex
// ---------------------------------------------------------------------------
// One hexagonal cell from the GPSJam daily dataset.
// Served via GET /api/gpsjam/current as a JSON array.

// JammingHex represents GPS interference data for one H3 hexagon on one day.
type JammingHex struct {
	H3Index         string  `json:"h3_index"`         // H3 hex identifier, e.g. "8928308280fffff"
	InterferencePct float64 `json:"interference_pct"` // 0.0–100.0
	Date            string  `json:"date"`             // "YYYY-MM-DD"
}

// ---------------------------------------------------------------------------
// Category B — TLERecord
// ---------------------------------------------------------------------------
// One satellite's Two-Line Element set, fetched from CelesTrak.
// Served via GET /api/satellites/tles as a JSON array.
// The browser propagates positions client-side using satellite.js.

// TLERecord holds the orbital data for a single satellite.
type TLERecord struct {
	NoradCatID    int       `json:"norad_cat_id"`
	Name          string    `json:"name"`
	TLELine1      string    `json:"tle_line1"`
	TLELine2      string    `json:"tle_line2"`
	Constellation string    `json:"constellation"` // "starlink" | "iridium-next" | "active" | ...
	FetchedAt     time.Time `json:"fetched_at"`
}
